### ISSUE TITLE
Ошибка разбора модуля

### ISSUE THREAD

**Issue by shcherbakov-vadim** [1C-Company/1c-edt-issues#29]
## Опишите ошибку

В модуле, содержащем области возникает ошибка разбора, из-за которой обычная работа с ним становится невозможной

## Как воспроизвести

1. Импортировать в новый проект конфигурацию 1С:ЕРПУХ 3.1.4.38
2. Открыть модуль формы документа "Списание безналичных денежных средств"

## Скриншоты

![2021-10-26_10-41-05](https://user-images.githubusercontent.com/52153345/139503934-6f47030f-7e93-4e47-ac0a-ef3f798169dd.png)

## Окружение:

 - Версия 1C:EDT 2021.2.6

**Comment by shcherbakov-vadim** [comment]
Еще один пример от разработчиков ERPУХ:
![Недопустимая лексема](https://user-images.githubusercontent.com/52153345/147250806-40414c6e-4aca-4968-868d-b7b6c96401af.png)

**Comment by marmyshev** [comment]
Вроде в документации Платформы написано, что так нельзя код писать (тегами препроцессора разбивать цельные синтаксические конструкции), отдельно для тех кто не хочет читать справку в Конфигураторе - написали Стандарт на ИТС что «так нельзя»…
Но воз и ныне там… 😭😭😭

**Comment by shcherbakov-vadim** [comment]
С точки зрения платформы это синтаксически корректный код, поэтому необходимо либо требовать его запрета на уровне платформы, либо добавить игнорирование областей, которые разбивают синтаксические конструкции.

**Comment by tretyakevich** [comment]
> С точки зрения платформы это синтаксически корректный код, поэтому необходимо либо требовать его запрета на уровне платформы, либо добавить игнорирование областей, которые разбивают синтаксические конструкции.

Да, мы обсудили этот момент. Сейчас это - ограничение реализации в EDT. Ограничение было внесено намеренно, поскольку сама платформа подобный код считает нерекомендуемым к использованию. К сожалению, это теория - а на практике, как оказывается - даже типовые конфигурации (судя по этому примеру) используют подобный код.

Проблему зафиксировали. К сожалению, пока быстрых (по времени) вариантов решения не нашли, будем думать.

**Comment by shcherbakov-vadim** [comment]
Вижу потенциальную проблему в расширениях при использовании аннотации ИзменениеИКонтроль - https://its.1c.ru/db/v8320doc#bookmark:dev:TI000002149. Там тоже вполне возможны такие конструкции, например если нужно заменить условие в предложении Если.

**Comment by marmyshev** [comment]
> С точки зрения платформы это синтаксически корректный код, поэтому необходимо либо требовать его запрета на уровне платформы, либо добавить игнорирование областей, которые разбивают синтаксические конструкции.

Что за выдумки?)) В какой момент это стало корректным кодом? Я что-то пропустил?))
Наличие проверки в Конфигураторе и "синтакисечки корректный код" - это разные вещи.  Вот создатели Платформы "с точки зрения платформы" считают такой код не корретным (см. статью в `Конфигураторе` в синтсксис-помощнике "`Инструкции препроцессора`" в версии 8.3.20) но не сдлеали для вас диагностику на такие ошибки:



> Поскольку единственное назначение инструкций препроцессора #Область и #КонецОбласти состоит в обозначении группируемых или группируемых и сворачиваемых строк в модулях на встроенном языке, не поддерживается пересечение этих инструкций с другими группируемыми конструкциями встроенного языка, например, с процедурами/функциями, условными операторами, циклами.

Правильно:

```bsl
#Область Обл1
…
Процедура Проц1()
…
КонецПроцедуры
…
#КонецОбласти
```

Неправильно (группируемые части модуля пересекаются):

```bsl
#Область Обл1
…
Процедура Проц1()
…
#КонецОбласти
…
КонецПроцедуры
```

**Comment by marmyshev** [comment]
> поэтому необходимо либо требовать его запрета на уровне платформы

Если отсутсвие диагностики  в Конфигураторе для вас автоматически делает такой код "коренным" - требуйте от Платформы этого запрета.

**Comment by shcherbakov-vadim** [comment]
Не нужно начитать демагогию. Есть проблема, что если в коде встречаются такие моменты синтаксический анализатор перестает работать, в отличие от конфигуратора. Вы можете закрыть задачу с формулировкой, что это корректное поведение и добавить новый пункт в раздел "Текущие ограничения функциональности".

**Comment by marmyshev** [comment]
1. Здесь нет демагогий, только пояснения. https://ru.wikipedia.org/wiki/Демагогия
2. Вы выдвигаете необоснованные предположения "правильности кода", тогда как Платформа явно говорит что код не корректный.
3. В Конфигураторе нет анализатора, в отличии от EDT (том плане что препроцессор в ЕДТ имеет аналитическое значение доступности кода на выбранном контексте).
4. Поведение проектное, т.е. это не ошибка. Это просто пояснение.
5. Это не ограничение функциональность - это реализация функциональности проверки корректности препроцессора, которой не было в Конфигураторе, но которая западина в документации Платформы.
6. Жалобы на это поведение всё равно есть - поэтому предлагаю не закрывать задачу. Вдруг, в будущем получится что-то улучшить...

**Comment by shcherbakov-vadim** [comment]
> 1. Здесь нет демагогий, только пояснения. [https://ru.wikipedia.org/wiki/Демагогия](https://ru.wikipedia.org/wiki/%D0%94%D0%B5%D0%BC%D0%B0%D0%B3%D0%BE%D0%B3%D0%B8%D1%8F)
Вы приводите собственную интерпретацию фразы "синтаксически корректный код", считая ее единственно верной и игнорируете более логичные варианты.
Вместо обсуждения проблемы вы начинаете диалог в стиле "сам дурак", при этом я даже не автор данного кода.
> 2. Вы выдвигаете необоснованные предположения "правильности кода", тогда как Платформа явно говорит что код не корректный.
В моем понимании "синтаксически корректный код" это то, что платформа работает без ошибок. То есть если бы платформа генерировала исключение такой код просто не появился бы.
> 3. В Конфигураторе нет анализатора, в отличии от EDT (том плане что препроцессор в ЕДТ имеет аналитическое значение доступности кода на выбранном контексте).
Как же тогда конфигуратор выполняет синтаксический контроль модулей с расширенной проверкой?
> 4. Поведение проектное, т.е. это не ошибка. Это просто пояснение.
Я выше привел пример, что подобная проблема может возникнуть с расширениями при использовании аннотации ИсправлениеИКонтроль. Сейчас провели и оно действительно так:
В конфигурации:
```bsl
Процедура Проверка1()
	Если 1 = 1 Тогда
		Сообщить(1);
	КонецЕсли;
КонецПроцедуры
```
В расширении:
```bsl
&ИзменениеИКонтроль("Проверка1")
Процедура Расш1_Проверка1()	
	#Удаление
	Если 1 = 1 Тогда
	#КонецУдаления
	#Вставка
	Если 2 = 2 Тогда
	#КонецВставки
		Сообщить(1);
	КонецЕсли;
КонецПроцедуры
```
В этом случае таких катастрофических последствий нет, но ошибки все равно выводятся. Это корректное поведение?
> 5. Это не ограничение функциональность - это реализация функциональности проверки корректности препроцессора, которой не было в Конфигураторе, но которая западина в документации Платформы.
Это существенное отличие поведение от конфигуратора, которое делает доработку модулей с такой проблемой практически невозможной.

**Comment by shcherbakov-vadim** [comment]
В документации по платформе понятие пересечение тоже можно трактовать по-разному. Например следующий фрагмент в EDT приводит к аналогичной ошибке:
```bsl
Процедура Проверка1()
	Если 1 = 1 Тогда
		Сообщить(1);
	#Область Область2
	ИначеЕсли 2 = 2 Тогда
		Сообщить(2);
	#КонецОбласти
	КонецЕсли;
КонецПроцедуры 
```
Однако в конфигураторе его свертка выполняется правильно:
![Свертка](https://user-images.githubusercontent.com/52153345/147500951-1851b9dd-a413-4e9f-a71b-8b5a99e85f02.png)

**Comment by Roffild** [comment]
https://github.com/1C-Company/1c-edt-issues/issues/1475

**Comment by JaneGlass73** [comment]
Сильно актуально в эту же тему, я за правильный код, но как бы "убедить" команду УХ, чтобы не делали таких вставок и поправили везде где есть уже. я им писала и не раз, но они меня игнорируют... В итоге воз и ныне там, и то что описано выше только верхушка айсберга, потому что как только в такие модули добавить процедуры и функции ломается попроцедурное сравнение и проблема переходит на следующий уровень...
https://github.com/1C-Company/1c-edt-issues/issues/1475#issuecomment-2564812112
